src/nupic/algorithms/BacktrackingTMCpp.cpp

```diff
@@ -802,72 +804,85 @@ std::shared_ptr<Real> BacktrackingTMCpp::getConfHistogram() {
-  deepcopyRestore_<Byte>(ss, "infActiveStateT", cells4_->getInfActiveStateT(),
+  deepcopyRestore_(ss, "infActiveStateT", cells4_->getInfActiveStateT(),
                          nCells);
-  deepcopyRestore_<Byte>(ss, "infActiveStateT1", cells4_->getInfActiveStateT1(),
+  deepcopyRestore_(ss, "infActiveStateT1", cells4_->getInfActiveStateT1(),
                          nCells);
-  deepcopyRestore_<Byte>(ss, "infPredictedStateT",
+  deepcopyRestore_(ss, "infPredictedStateT",
                          cells4_->getInfPredictedStateT(), nCells);
-  deepcopyRestore_<Byte>(ss, "infPredictedStateT1",
+  deepcopyRestore_(ss, "infPredictedStateT1",
                          cells4_->getInfPredictedStateT1(), nCells);
-  deepcopyRestore_<Byte>(ss, "learnActiveStateT",
+  deepcopyRestore_(ss, "learnActiveStateT",
                          cells4_->getLearnActiveStateT(), nCells);
-  deepcopyRestore_<Byte>(ss, "learnActiveStateT1",
+  deepcopyRestore_(ss, "learnActiveStateT1",
                          cells4_->getLearnActiveStateT1(), nCells);
-  deepcopyRestore_<Byte>(ss, "learnPredictedStateT",
+  deepcopyRestore_(ss, "learnPredictedStateT",
                          cells4_->getLearnPredictedStateT(), nCells);
-  deepcopyRestore_<Byte>(ss, "learnPredictedStateT1",
+  deepcopyRestore_(ss, "learnPredictedStateT1",
                          cells4_->getLearnPredictedStateT1(), nCells);
-  deepcopyRestore_<Real>(ss, "cellConfidenceT", cells4_->getCellConfidenceT(),
+  deepcopyRestore_(ss, "cellConfidenceT", cells4_->getCellConfidenceT(),
                          nCells);
-  deepcopyRestore_<Real>(ss, "cellConfidenceT1", cells4_->getCellConfidenceT1(),
+  deepcopyRestore_(ss, "cellConfidenceT1", cells4_->getCellConfidenceT1(),
                          nCells);
-  deepcopyRestore_<Real>(ss, "colConfidenceT", cells4_->getColConfidenceT(),
+  deepcopyRestore_(ss, "colConfidenceT", cells4_->getColConfidenceT(),
                          loc_.numberOfCols);
-  deepcopyRestore_<Real>(ss, "colConfidenceT", cells4_->getColConfidenceT(),
+  deepcopyRestore_(ss, "colConfidenceT", cells4_->getColConfidenceT(),
                          loc_.numberOfCols);
 }
 
-template <typename T>
 void BacktrackingTMCpp::deepcopySave_(tmSavedState_t &ss, std::string name,
-                                      T *buf, Size count) {
-  std::shared_ptr<T> target(new T[count], std::default_delete<T[]>());
-  fastbuffercopy<T>(target.get(), buf, count);
-  ss[name] = target; // 'target' could be a shared_ptr<Byte> or shared_ptr<Real>
+                                      Byte *buf, Size count) {
+  std::shared_ptr<Byte> target(new Byte[count], std::default_delete<Byte[]>());
+  fastbuffercopy<Byte>(target.get(), buf, count);
+  struct ss_t val;
+  val.Byteptr = target;
+  ss[name] = val;
+}
+void BacktrackingTMCpp::deepcopySave_(tmSavedState_t &ss, std::string name,
+                                      Real *buf, Size count) {
```

Comment:
keep the templated method? Or make a private templated; and overload with "public"/protected Byte/Real variants?